### ISSUE TITLE
How to do Inference on multiple images

### ISSUE THREAD

**Issue by Ram124** [ijkguo/mx-rcnn#95]
@ijkguo @li-haoran 
How to get inference for multiple images with the help of demo.py?
I have 2 classes. i have done training.. I have got the test accuracy of 85%.
Now i want to visualize the output on some 100's of images. 
How to proceed with it?
because with demo.py we could visualize only 1 image at a time..
How to get it for N number of images ...

Any help is really appreciated.

**Comment by ijkguo** [comment]
Follow the demo script and use a for loop to do inference on a list of images.

**Comment by Ram124** [comment]
@ijkguo I have done training on 200 images as example.
But i want to test it on some images.
But when i run demo.py. i am getting only 1 class per image.
I have total of 3 classes ( person, bicycle, background)
i have feeded the network with 101 images of each class.

What could be the problem?

**Comment by ijkguo** [comment]
Have you resolve the original issue, do demo on multiple images?

**Comment by Ram124** [comment]
@ijkguo Yes i have done...

**Comment by ijkguo** [comment]
Then your description of the new problem is that you have 200 images, 101 images contain person, 101 contain bicycle and 101 contain background (so 3 of them has all 3 classes?).
After training (not clear how to feed 101 of image to network?), you run demo scripts on (some data, not known but could have person and bicycle?).
The output (not known, maybe visualization or printout?) is only 1 class per image.

**Comment by Ram124** [comment]
No, 
1. I have 2 classes ( person, Bicycle)
2. I have taken 100 images with 1 person in each image. And 100 images with 1 bicycle in each image. (All single object images)
3. I have feeded these to the network (total 200 images)

4. After training, I executed test.py. 
And i got 80% accuracy.
5. Now i want to check this Model on Some fresh data . This fresh data contains both Person and Bicycle in Each image (dual object images)
This model has to predict the classes correctly..

But i am getting only 1 object detected per image..

But i trained the netowork using incubator mxnet-rcnn and got the model after 8 epochs. 
When i give that params file to demo.py in this repo, i could able to detect the classes (not so great, but it is okay. Some images are detected correctly)

Now I want to increase the dataset and do it again..

**Comment by ijkguo** [comment]
Make sure all config (printed or in code) between mxnet/rcnn/example and mx-rcnn are the same. Fun facts: VOC07++12 has 16k images; COCO train2017 has 118k images; KITTI trainval has 7k images. Look at some sample images from COCO in the README: they have many objects in one image.

**Comment by Ram124** [comment]
@ijkguo  I have checked them.  I found 
network.resnet.FIXED_PARAMS_SHARED = ['conv0', 'stage1', 'stage2', 'stage3', 'gamma', 'beta']  
config.TRAIN.RPN_CLOBBER_POSITIVES = False
# rpn bounding box regression params
config.TRAIN.RPN_BBOX_WEIGHTS = (1.0, 1.0, 1.0, 1.0)
config.TRAIN.RPN_POSITIVE_WEIGHT = -1.0

config.TRAIN.BG_THRESH_HI = 0.5
config.TRAIN.BG_THRESH_LO = 0.0
# rcnn bounding box regression params
config.TRAIN.BBOX_REGRESSION_THRESH = 0.5
config.TRAIN.BBOX_WEIGHTS = np.array([1.0, 1.0, 1.0, 1.0])

These configurations are not given anywhere..
Is that the problem for not getting the right param file after training is done?

**Comment by Ram124** [comment]
I have checked the parameters in this repo with the incubator/(rcnn example.. when training script is made to run,  There are some difference. 
bbox_target, bbox_weight has different size
rpn_bbox_loss_output
rpn_cls_prob_output are different.

what is the problem.??


INFO:root:max input shape
{'bbox_target': (1, 36, 38, 63),
 'bbox_weight': (1, 36, 38, 63),
 'data': (1, 3, 600, 1000),
 'gt_boxes': (1, 100, 5),
 'im_info': (1, 3),
 'label': (1, 1, 342, 63)}
INFO:root:max output shape
{'bbox_loss_reshape_output': (1, 128, 12),
 'blockgrad0_output': (1, 128),
 'cls_prob_reshape_output': (1, 128, 3),
 'rpn_bbox_loss_output': (1, 36, 38, 63),
 'rpn_cls_prob_output': (1, 2, 342, 63)}


INFO:root:providing maximum shape 
[('data', (1, 3, 600, 1000)), 
('gt_boxes', (1, 100, 5))] 
[('label', (1, 20646)), 
('bbox_target', (1, 36, 37, 62)), 
('bbox_weight', (1, 36, 37, 62))]
INFO:root:output shape {'bbox_loss_reshape_output': (1, 128, 12),
 'blockgrad0_output': (1, 128),
 'cls_prob_reshape_output': (1, 128, 3),
 'rpn_bbox_loss_output': (1, 36, 37, 37),
 'rpn_cls_prob_output': (1, 2, 333, 37)}

**Comment by Ram124** [comment]
@ijkguo 
My input images are grayscale  images. But vgg16 takes only rgb input. 
Can i feed the network with grayscale images (1, 512, 512) instead of (3, 512, 512)?

In this repo, the input is fed as rgb irrespective of the input channels..
Is there any way make it SINGLE CHANNEL INPUT?

**Comment by ijkguo** [comment]
Our models are initialized from models trained on ImageNet image classification dataset, which are color images. You might want a pretrained model on grey scale images.

**Comment by Ram124** [comment]
Is there a pretrained model on grey scale images compatible with mxnet? 
Is mobilenet accepts grey scale images??
Can i use those convolution layers  instead of vgg16 and resnet layers..??

**Comment by ijkguo** [comment]
I am not aware of grey scale models.
